memory/src/main/scala/filodb.memory/format/MemoryReader.scala

```diff
@@ -0,0 +1,163 @@
+  final def getLongVolatile(addr: Long): Long = unsafe.getLongVolatile(base, baseOffset + addr)
+  final def getDouble(addr: Long): Double = unsafe.getDouble(base, baseOffset + addr)
+  final def getFloat(addr: Long): Double = unsafe.getFloat(base, baseOffset + addr)
+}
+
+/**
+  * Factory methods for Memory Accessors
+  */
+object MemoryAccessor {
+  def fromArray(array: Array[Byte]): MemoryAccessor = new ByteArrayAccessor(array)
+  def nativePtrAccessor: MemoryAccessor = NativePointerAccessor
+  def fromByteBuffer(buf: ByteBuffer): MemoryAccessor = {
+    if (buf.isDirect) DirectBufferAccessor(buf)
+    else new OnHeapByteBufferAccessor(buf)
+  }
+}
+
+/**
+  * Abstraction for reading and writing to memory byte sequence. Clients can used this
+  * abstraction without worrying whether underlying memory could be
+  * on-heap or off-heap.
+  *
+  * The `addr` parameter in the methods below indicate the address of the
+  * memory position relative to the start of the byte sequence represented by the
+  * implementing class. IMPORTANT to note that the addr parameter is NOT same as the unsafe
+  * offset for any given on-heap object.
+  */
+sealed trait MemoryAccessor extends MemoryReader {
+  final def setByte(addr: Long, byt: Byte): Unit = unsafe.putByte(base, baseOffset + addr, byt)
+  final def setShort(addr: Long, s: Short): Unit = unsafe.putShort(base, baseOffset + addr, s)
+  final def setInt(addr: Long, i: Int): Unit = unsafe.putInt(base, baseOffset + addr, i)
+  final def setIntVolatile(addr: Long, i: Int): Unit = unsafe.putIntVolatile(base, baseOffset + addr, i)
+  final def setLong(addr: Long, l: Long): Unit = unsafe.putLong(base, baseOffset + addr, l)
+  final def setDouble(addr: Long, d: Double): Unit = unsafe.putDouble(base, baseOffset + addr, d)
+  final def setFloat(addr: Long, f: Float): Unit = unsafe.putFloat(base, baseOffset + addr, f)
+  final def copy(fromAddr: Long, toAcc: MemoryReader, toAddr: Long, numBytes: Int): Unit =
+    UnsafeUtils.copy(base, baseOffset + fromAddr, toAcc.base, toAcc.baseOffset + toAddr, numBytes)
+  //  def wordCompare(thisOffset: Long, destObj: MemoryBase, destOffset: Long, n: Int): Int
+  //  def compareTo(offset1: Long, numBytes1: Int, base2: MemoryBase, offset2: Long, numBytes2: Int): Int
+}
+
+/**
+  * Implementation used to access a native pointer. The `addr` parameter in methods imply
+  * a raw native pointer.
+  *
+  * This is a static implementation of MemoryReader to avoid allocation
+  * per pointer.
+  *
+  * One could envision an alternate implementation which pays the cost of an on-heap
+  * allocation but additionally provides bounds checks for safe memory access.
+  */
+object NativePointerAccessor extends MemoryAccessor {
+  // TODO check bounds of array before accessing
+  val base = ZeroPointer
+  val baseOffset: Long = 0
+
+  override def wrapInto(buf: DirectBuffer, addr: Long, length: Int): Unit = {
+    buf.wrap(addr, length)
+  }
+}
```

Comment:
Clarification of terms here. In my mind, offset is the unsafe offset. Let us agree not to use the term offset in any other context. There is a barrier in communication because of these terms.

If you are doing byte position arithmetic, that is an acceptable thing both on native pointers as well as accessor addresses. I don't see any violation of abstraction here. However, arithmetic with unsafe offset for on-heap should not be done by API clients. I have called out a TODO to make base and baseOffset APIs to be qualified as "protected" to prevent this. This can be done in future iterations.

Reply:
We decided to defer this effort to next iteration.